### ISSUE TITLE
Worker

### ISSUE THREAD

**Issue by droyad** [OctopusDeploy/Specs#23]
This component runs `Run on server` steps and script console scripts.

**Comment by droyad** [comment]
WIP

**Comment by MJRichardson** [comment]
I feel the same.

It will be interesting to see how much demand for point 2 (hosts with different capabilities) there is from hosted customers, because they obviously can't control the server capabilities as they can on-prem.

**Comment by matt-richardson** [comment]
> but makes it much more complicated.
Can you expand on this?

re
> Isolation from other deployments and the database (both for security and performance)

I'd like to have a discussion about what, if any, benefit exists in having a security boundary between the orchestrator and the worker. I don't think I've got a clear enough picture at the moment.

From an SOA point of view, feels like orchestrating and running stuff are separate responsibilities. 

re 
> Ability to have hosts with different capabilities (eg Linux, Azure SDK)
Within a few weeks of deploying Octopus as a customer, I ran into a need to use different Azure SDK versions for different teams. While there was a workaround in using a single Tentacle for that, we all felt it was icky, as it was a single point of failure, and that it was a workaround rather than something supported well.

One of the gains I see with supporting different hosts is that for hosted we can default to running on Windows Nano (a 92 meg, super fast launching container) rather than Windows Server Core (an 8GB, slow running container), which will give us benefits in density (therefore reducing cost), and speed.

It also gives customers a lot of benefits in being able to run things like (to pick something random) a security scanning container super easy. 

I think there are also compelling advantages here with changing the perception that Octopus is a .net only tool. If we just are spinning up containers, they could be any OS which would be more expected from people from other languages.

**Comment by droyad** [comment]
As mentioned in https://github.com/OctopusDeploy/Specs/issues/21#issuecomment-345648489 I now think we should do this.

I also realised that the Worker is just a tentacle, it does exactly the same thing.

**Comment by droyad** [comment]
@matt-richardson 
> but makes it more complicated
We would have to introduce a mechanism for a `Run on Server` step to choose which pool of workers the step should run on (eg  via Roles, Prerequisites or Tags).

> I'd like to have a discussion about what, if any, benefit exists in having a security boundary between the orchestrator and the worker
In the initial design not so much, that's why I was thinking to keep it within the orchestrator. However if we combine the orchestator and TaskRunner (See https://github.com/OctopusDeploy/Specs/issues/21#issuecomment-345648489) we definitely want a security boundry between the Task Runner and Worker

> From an SOA point of view, feels like orchestrating and running stuff are separate responsibilities.
Are you arguing for or against the worker being on a separate system? I don't think we can get to hung up on architectural purity at the expense of installability/maintainability for customers.

Also, I believe that it's possible to have SoC even if it is within the same process. When I am talking about "seperate thing", I mean it should be possible to run it on a different machine. Internally they are, and will continue to be different components within the overall system.

> Within a few weeks of deploying Octopus as a customer, I ran into a need to use different Azure SDK versions for different teams
As I said, it would be a nice to have. Have can't do everything at once.

**Comment by MichaelJCompton** [comment]
## Here's an initial cut of workers stage 1:

This is the how to assign workers and run part.  The idea is that worker pools, dynamic provisioning, capabilities etc, can happen after (and a bit independent of) this first cut.

- Workers are just tentacles.
- run on server steps become 'run once'
- A worker is tagged with `worker in environments : Env A, Env C` etc, same as we tag machines to environments.
- We assume if you tag a machine as a worker, then it's got everything _you_ could need for a worker in that environment - but we don't care what that is (e.g. Matt's comment above: if you want different versions of Azure SDK in different environments, go for it, we won't even check.  But the deployment will fall over if you don't set it up right).
- We allow 'pools' of these workers in any environment.  At deploy time, when we hit a 'run once' or 'run once for role' we do a similar search as now for applicable machines, but non-deterministically pick 1 machine that matches.  (As per point above, if there are multiple workers in an env, we are assuming that they are essentially identical)
- The running of the worker task is thus no different to running any other task
- Because workers are tentacles, health checks, being disabled, etc all happen in the existing way.  Can also carry this over to permissions, so the machine permissions also imply setting up workers
- For small installs it works out of the box on the server and no-one has to think about it.

So this allows for example to just use deployment targets in your environment as workers (take the coloured boxes as has been ticked as a worker for that env)

![image](https://user-images.githubusercontent.com/13693742/33423053-cdce7406-d60b-11e7-9653-2d6a6d271d59.png)

Or you could assign a worker pool outside of the environments

![image](https://user-images.githubusercontent.com/13693742/33423093-ed41331e-d60b-11e7-9ebf-bb9837c0474b.png)

Or say just reuse targets as workers in Dev, but have a dedicated pool of workers in prod

![image](https://user-images.githubusercontent.com/13693742/33423142-169debf8-d60c-11e7-9dd7-0259f58c222e.png)

Allowing targets to be workers (if that's how a user chose to set it up) means the machine wide mutex would block a target from doing a worker task and a deployment task simultaneously - so that's safe and would only crop up if a deployment was running run once and deployment steps in parallel or the machine was involved in multiple deployments (say as a target in one and a worker in another).

Having a pool of workers allows a number of run once steps to happen in parallel.

For the first cut of hosted this could remove any requirement on us to provide workers or do run once tasks at our cost at all.

## UI

For the UI it's small changes:

- on a machine's settings page it needs options for ticking it as a worker (same as setting it to an environment)

- on the infrastructure page, a box for workers by environment

![image](https://user-images.githubusercontent.com/13693742/33423749-dff9c9a8-d60d-11e7-9de5-c74c5bbb218f.png)

- on a environments page, we could tag machines that are workers (but this might be confusing as they could be working in another environment)

![image](https://user-images.githubusercontent.com/13693742/33423858-1f2f6646-d60e-11e7-9aab-445c7ea0f192.png)


## Small installs
 
Having an out of the box working version for small installs needs a tentacle installed on the server and something like a global flag `has default worker = machineABC` or a tag on that machine that causes the get applicable machines function to pick the default machine for all run once tasks in all environments.

Once there's a need to scale workers, a user can untick the option and start provisioning workers.

![image](https://user-images.githubusercontent.com/13693742/33424146-ff73c788-d60e-11e7-8785-10a28a1de7a0.png)

## Later

After round 1 then we can think about

- maybe tag workers with role+environment
- scaling / provisioning workers can happen outside the mechanics above : the above just asks 'get me an applicable worker for this environment/role' and doesn't care if the machine was spun up just for this task or if some smart scheduling was done to pick an idle worker etc.
- worker capabilities : there's loads of options here (maybe we need to throw stage 1 over the fence and see what people ask for).  Do we want to detect a known set of capabilities in health checks, do we want to allow users to tag machines with their own set of flags ?  either way large parts of the above are resilient to these choices and most of the complexity still gets push into the 'pick me a machine with these constraints' function.

**Comment by matt-richardson** [comment]
A few thoughts:
> At deploy time, when we hit a 'run once' or 'run once for role' we do a similar search as now for applicable machines, but non-deterministically pick 1 machine that matches.
We'd need to filter out the ones that are already running something, and probably deal with the issue of none available, so choose the first one that becomes available, rather than the pick it at schedule time. (Otherwise, we'd potentially choose the worker that's going to take 3 hours to finish its task, when another one becomes available after a few seconds.)
We'd probably also want to try and avoid choosing a worker that is also a target in the same deployment if possible (ie, by ordering the list so that these are at the bottom), so we can limit the time impact.

> For the first cut of hosted this could remove any requirement on us to provide workers or do run once tasks at our cost at all.
There's been a very strong preference stated that we support "run once" tasks without any customer provided infrastructure. (It would solve a whole bunch of problems if its customer provided though.)

It would be very useful to think about exposing worker usage metrics fairly early on with this, so we can trigger scaling mechanisms.

**Comment by zentron** [comment]
Adding as a reference open UserVoice ticket
https://octopusdeploy.uservoice.com/forums/170787-general/suggestions/3721946-ability-to-run-a-powershell-script-once-before

Note the additional comment regarding `run N times`

**Comment by MichaelJCompton** [comment]
# Options for specifying workers.

I feel like there’s three parts to workers: 

1) Ability to specify workers, and thus add a run-once step
2) During a deployment pick a worker
3) Orchestrate tasks on workers

At the moment (with run on server) those steps look like 

1_now) Tick as run on server
2_now) Switch (in “task runner”) between it’s a deployment task on a tentacle or a run on server step
3_now) Invoke calamari locally and run the script

It feels like once we allow tentacle workers that (3) should just look like the steps for running on a tentacle and (2) is a function that picks an appropriate worker given the kinds of constraints in (1) - (this could be simply find a machine that satisfies the constraints, or as Matt points out this could include scheduling concerns etc).  Those look like they fall out after settling on (1), so to my mind the question about workers is how to allow specifying something as a worker.

A nice endpoint would be workers and targets in this kind of model

https://github.com/OctopusDeploy/Specs/blob/master/Resources/index.md

Until we get there what’s a sensible way to identify workers - here’s some options


## Flag as a worker for an environment

Machines have a property e.g. `ENV_A | ENV_B | …`, which, when set, makes a machines a worker in the given environments.  For run once steps, we pick any machine marked as a worker for the current deployment environment.

Allows

- deployment targets to double up as spare workers
- also workers that aren’t targets
- worker pools per environment
- dedicated worker machines
- different workers in different environments

But doesn’t allow worker capabilities.

## Flag with worker + tag/role

Machines are marked for worker in the same way as we mark deployment targets - a list of environments and a list of tags/roles.  The tags allow setting capabilities etc. for workers, e.g `net core`, `DB worker`, `linux worker`, etc.   For run once steps we allow something like “run once (on a machine tagged with <tag>)” and “run once for each <role>  (on a machine tagged with <tag>)”.   Maybe make the tags optional and it would pick any worker.

Allows

- the above, plus
- worker capabilities
- can pick out deployment machines as workers - “I really want this to run once on one of my DB targets, but I don’t care which”


So the tags act like names for worker pools - but from a user perspective could mean pools of capabilities (though we aren’t going to check or detect those).

As noted over [here](https://github.com/OctopusDeploy/Specs/issues/30) this option is pretty much what we already do with selecting deployment targets, so we already have the machinery to pick all the machines with an ENV+tag/role combination, so we already have all the pieces to pick appropriate workers.

**Comment by MichaelJCompton** [comment]
Upgrades / installs

- should we leave the capability to run in process (in task runner) so there’s no friction for upgrades/installs, or do we split it completely out and thus require a tentacle (on the server, or elsewhere) as a default worker.

- if we do split out then the upgrade path needs to include an option/advice about adding a default worker (or many workers)

- similarly installs need to have a sensible default - the server install would need to do something like install and setup both a server and a tentacle

Maybe it’s better not to split completely out in the first instance, leave the capability to run on server, but add the options for workers?  Internally, then if no worker annotations are picked up it looks just like now, if worker annotations are picked up it runs on one such worker.  That would be the least change - to customers it looks like a feature addition not cheese moving.

**Comment by zentron** [comment]
I think for migration we could just "install a worker" on the same machine as the server so that they are none the wiser?

**Comment by MichaelJCompton** [comment]
That's installing and tentacle, opening a port, etc. would they be none the wiser?

**Comment by MichaelJCompton** [comment]
Slides for TL;DR 

[TlDR_Workers.pptx](https://github.com/OctopusDeploy/Specs/files/1546277/TlDR_Workers.pptx)

**Comment by MichaelJCompton** [comment]
more user voice tickets that relate to this (not all recent)

https://octopusdeploy.uservoice.com/forums/170787-general/suggestions/20157787-make-run-on-octopus-server-a-separate-permission

https://octopusdeploy.uservoice.com/forums/170787-general/suggestions/10172955-allow-azure-powershell-steps-to-run-on-servers-oth

https://octopusdeploy.uservoice.com/forums/170787-general/suggestions/6316906-support-run-on-any-tentacle-model-for-deployment
